### PR TITLE
[WIP] Add vanilla Neural ODE module

### PR DESCRIPTION
**Pull request by Zymrael**
## What does this PR do?

This PR sets the stage for future integration with continuous-depth models and neural differential equations ([here](https://github.com/Zymrael/awesome-neural-ode) for a brief overview of what's happening in the space)

We've discussed contributing with neural diff. equation models directly on Slack. This PR shows a possible way in which these models could be integrated in `bolts`, relying on [torchdyn](https://github.com/DiffEqML/torchdyn). 

The main question is of course whether we'd be okay with relying on an external library (`torchdyn`). If this is determined to be a non-issue there are a lot more features and modules that can be added fairly easily, such as Neural ODEs for image classification and continuous normalizing flows.

Some things are missing (`main_cli`, a few tests, more variants) -- will be added.

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?
- [ ] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/CHANGELOG.md)?

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Sure 😄

### CODE REVIEW

**Comment by Borda** [comment]
@Zymrael nice, could you please add requirement here: https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/requirements/models.txt

**Comment by Zymrael** [comment]
@Borda Cool, I see how you've split the dependencies. I'll get back to this PR after our (very close) next pip release of `torchdyn`

**Comment by Borda** [comment]
@Zymrael how is it going here, is it still WIP? 🐰

**Comment by Borda** [comment]
@Zymrael any update here? 👼

**Comment by Zymrael** [comment]
@Borda I'll find some time this week to finish preparing this PR, been busy with conference deadlines :)

**Comment by Zymrael** [comment]
We have the test and the full CI test suite seems to be passing (still running at the time where I left this comment). Other workflows do not seem to be liking the `torchdyn` dependency in `requirements/models`. 

```
You should consider upgrading via the '/opt/hostedtoolcache/Python/3.7.9/x64/bin/python -m pip install --upgrade pip' command.
error in pytorch-lightning-bolts setup command: 'extras_require' must be a dictionary whose values are strings or lists of strings containing valid project/version requirement specifiers.
Partial import of `pl_bolts` during the build process.
```
The problems seem to be unrelated though, feedback?

File: pl_bolts/models/diffeq/neuralode.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
+from argparse import ArgumentParser
+
+import pytorch_lightning as pl
+import torch
+import torch.nn as nn
+from torch import Tensor
+
+from torchdyn.models import NeuralDE, DataControl
+
```
Comment (akihironitta) [inline pl_bolts/models/diffeq/neuralode.py:9]: We need to import optional packages as the following to avoid import errors when optional packages are not installed. I think this should solve the import errors in the doctest.
```suggestion

from pl_bolts.utils import _TORCHDYN_AVAILABLE
from pl_bolts.utils.warnings import warn_missing_pkg

if _TORCHDYN_AVAILABLE:
    from torchdyn.models import NeuralDE, DataControl
else:  # pragma: no cover
    warn_missing_pkg("torchdyn")
```

In addition to this suggestion, could you also define `_TORCHDYN_AVAILABLE` in `pl_bolts/utils/__init__.py`.

File: pl_bolts/models/diffeq/neuralode.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+            input_dim: size of input data
+            hidden_dim: size of hidden state
+            absolute_tol: absolute tolerance of the numerical solver
+            relative_tol: relative tolerance of the numerical solver
+            integration_span: mesh grid of evaluation points for the ODE
+            data_control: data control mechanism as described in https://arxiv.org/abs/2002.08071
+            learning_rate: learning rate for Adam
+
+        """
+        super().__init__()
```
Comment (akihironitta) [inline pl_bolts/models/diffeq/neuralode.py:45]: Let's raise an exception here with a helpful message for users.
```suggestion
        """
        if not _TORCHDYN_AVAILABLE:  # pragma: no cover
            raise ModuleNotFoundError("You want to use `torchdyn` which is not installed yet.")

        super().__init__()
```

File: pl_bolts/models/diffeq/neuralode.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+def cli_main(args=None):
+    from torchdyn.datasets import ToyDataset, generate_concentric_spheres, generate_moons
+    pl.seed_everything(1234)
+
+    parser = ArgumentParser()
+    parser.add_argument("--dataset", default="concentric_circles", type=str, help="concentric_circles, moons")
+    script_args, _ = parser.parse_known_args(args)
+
+    return None, None, None
```
Comment (akihironitta) [inline pl_bolts/models/diffeq/neuralode.py:111]: Why do we have these returns from `cli_main()`?

**Review by akihironitta — COMMENTED** [review]
@Zymrael Thank you for your work! Would you mind having a look at my suggestions?

**Comment by akihironitta** [comment]
Please feel free to reopen this PR.
